src/shell/shell.zig

```diff
@@ -816,12 +815,53 @@ pub const AST = struct {
             for (this.name_and_args) |*atom| {
                 cost += atom.memoryCost();
             }
+            cost += this.redirects.memoryCost();
+            return cost;
+        }
+    };
 
-            if (this.redirect_file) |*redirect_file| {
-                cost += redirect_file.memoryCost();
-            }
+    /// Redirect destinations for stdin, stdout, and stderr
+    pub const Redirects = struct {
+        stdin: ?RedirectTarget = null,
+        stdout: ?RedirectTarget = null,
+        stderr: ?RedirectTarget = null,
+
+        /// Type for specifying which IO stream
+        pub const IoKind = enum { stdin, stdout, stderr };
+
+        pub fn memoryCost(this: *const @This()) usize {
+            var cost: usize = @sizeOf(Redirects);
+            if (this.stdin) |*r| cost += r.memoryCost();
+            if (this.stdout) |*r| cost += r.memoryCost();
+            if (this.stderr) |*r| cost += r.memoryCost();
             return cost;
         }
+
+        /// Check if a given io stream is redirected elsewhere
+        pub fn redirectsElsewhere(this: *const Redirects, comptime io_kind: IoKind) bool {
+            return @field(this, @tagName(io_kind)) != null;
+        }
+    };
+
+    /// Where a stream is redirected to
+    pub const RedirectTarget = union(enum) {
+        /// Redirect to a file path (from shell text)
+        atom: struct {
+            atom: Atom,
+            append: bool = false,
+        },
+        /// Redirect to a JS object (ArrayBuffer, Blob, etc.)
+        jsbuf: JSBuf,
+        /// Duplicate to another fd (e.g., 2>&1 means stderr duplicates to stdout's destination)
+        dup: enum { stdin, stdout, stderr },
+
```

Comment:
Remove ?RedirectKind and use none: void instead so we avoid the struct size increase unnecessarily